### ISSUE TITLE
Chemacs 2

### ISSUE THREAD

**Issue by JacekHoleczek** [abougouffa/minemacs#4]
I am trying to use the [Chemacs 2](https://github.com/plexus/chemacs2) Emacs profile switcher with the [MinEmacs](https://github.com/abougouffa/minemacs) (and an up-to-date [emacs-28](https://git.savannah.gnu.org/cgit/emacs.git/log/?h=emacs-28) branch, built from source on a Ubuntu 22.04 / x86_64 machine).

I run:
```bash
rm -rf ~/.minemacs.d && \
mkdir ~/.minemacs.d && \
git clone --depth 1 https://github.com/abougouffa/minemacs.git ~/.minemacs.d/.emacs.d
```

I created a [Chemacs 2](https://github.com/plexus/chemacs2) profile (in my "`~/.emacs-profiles.el`" file):
```elisp
(
 ("minemacs" . ((user-emacs-directory . "~/.minemacs.d/.emacs.d/")
               (server-name . "minemacs")
               (env . (("MINEMACS_DIR" . "~/.minemacs.d/")))))
)
```

The problem is that it completely fails:
```bash
emacs --with-profile minemacs
```

I also tried a more brutal "bootstrapping" approach, but it failed, too:
```bash
echo '(load-file "~/.minemacs.d/.emacs.d/init.el")' > ~/.emacs.el
MINEMACS_DIR="${HOME}/.minemacs.d/" emacs
```

**Comment by abougouffa** [comment]
Hello @JacekHoleczek

Thanks for bringing this to my attention,

I've pushed a commit which hopefully will fix this. I've tested it with the same steps you've described.

Can you confirm if it works as expected?

**Comment by JacekHoleczek** [comment]
A small update ...

I decided to try it from scratch again (i.e., I removed the "`~/.minemacs.d`").
This time, it asked to download the "`Emoji images`" just once and there was no "timer" error.

Well, I forgot to report something slightly worrying ...
When installing, the [MinEmacs](https://github.com/abougouffa/minemacs) creates an empty "`.emacs.d/snippets`" directory (well, it should NOT touch "`~/.emacs.d`" at all).

**Comment by abougouffa** [comment]
Hello @JacekHoleczek, thank you for the feedback!

I'm using MinEmacs with Emacs 30 (built from source), the `emojify` config is used only on version 28, as Emacs 29+ have native emoji support (not configured perfectly ATM).

I've just pushed a commit to suppress the annoying `emojify` confirmation, I guess this could fix the timer error too, as `emojify` is lazy loaded at startup (which loads packages inside `run-with-idle-timer`).

For the brutal bootstrapping approach, I think there is a small issue with the current Chemacs2 implementation. When I tested it a VM, I got this:

```
Warning (initialization): An error occurred while loading ‘/home/vm/.emacs.d/early-init.el’:

error: [chemacs] ~/.emacs-profiles.el does not exist.

To ensure normal operation, you should investigate and remove the
cause of the error in your initialization file.  Start Emacs with
the ‘--debug-init’ option to view a complete error backtrace. Disable showing Disable logging
Warning (initialization): An error occurred while loading ‘/home/vm/.emacs.el’:

File is missing: Cannot open load file, No such file or directory, ~/.emacs.d/core/me-vars.el

To ensure normal operation, you should investigate and remove the
cause of the error in your initialization file.  Start Emacs with
the ‘--debug-init’ option to view a complete error backtrace. Disable showing Disable logging
```

This seems to be a problem with Chemacs2. Lets me explain: because `~/.emacs.d/early-init.el` exists, Emacs will first set `user-emacs-directory` to `~/.emacs.d/` and load `~/.emacs.d/early-init.el`, and then, **instead of loading `~/.emacs.d/init.el` it loads the prioritized file `~/.emacs.el`**.

This seems to me a bit awkward (in my machine, I removed `~/.emacs-profiles.el` and got the error above). The last commits should make Emacs happy (loads correctly without `me-vars.el` issue), however, loading Emacs this way it will miss the options set in MinEmacs' `early-init.el` (because it will load Chemacs' one instead).

I agree on the `MINEMACSDIR` part, now both `MINEMACSDIR` and `MINEMACS_DIR` are valid env vars, I kept the old one for backward compatibility.

**Comment by abougouffa** [comment]
The issue with snippets directory should be fixed also (I had other directories showing around like `transient/` and `pcache/`). Although, I think you will get occasionally some random directories in the `.emacs.d/`, because before, I was setting `user-emacs-directory` explicitly to `minemacs-local-dir`, but it seems that Chemacs2 is not OK with it!

**Comment by abougouffa** [comment]
@JacekHoleczek Can you try to pull the latest commits and try again?

**Comment by JacekHoleczek** [comment]
Great!

The download the "`Emoji images`" confirmation is gone.

It seems that the "`~/.emacs.d/snippets`" directory is always created, when the "`yasnippet-snippets`" or the "`doom-snippets`" are being worked on, regarless of the "bootstrapping" approach.

One gets new errors now, whenever one runs emacs (which maybe suggest that the above problem is related to "`yasnippet-...`", not "`doom-...`"):

```text
[yas] Check your `yas-snippet-dirs': ~/.minemacs.d/snippets/ is not a directory [3 times]
[yas] Prepared just-in-time loading of snippets with some errors.  Check *Messages*.
```

Except for the above problems, I can confirm that the [Chemacs 2](https://github.com/plexus/chemacs2) approach works now.

When trying the brutal "bootstrapping" approach ... I forgot that it will unconditionally execute the "`~/.emacs.d/early-init.el`" before the "`~/.emacs.el`".
The solution is simple.
Either (it seems that the "`-q`" option is not needed):

`MINEMACSDIR="${HOME}/.minemacs.d/" emacs --eval='(load-file "~/.minemacs.d/.emacs.d/init.el")'`

or:

```lisp
mv ~/.emacs.d ~/.emacs.d.hidden # we need to "hide" it
echo '(load-file "~/.minemacs.d/.emacs.d/init.el")' > ~/.emacs.el
MINEMACSDIR="${HOME}/.minemacs.d/" emacs
```

**Comment by abougouffa** [comment]
Thanks for the confirmation @JacekHoleczek

I guess the last commit should fix the snippets' error (by ensuring the directory exists, and create it if not).

**Comment by JacekHoleczek** [comment]
Almost perfect.

It still creates the spurious "`~/.emacs.d/snippets`" directory.

And it now also creates the "`~/.minemacs.d/snippets`" directory, so the error is gone.
Is this directory supposed to "replace" the original "`~/.emacs.d/snippets`" directory?
If yes, maybe it should create "`~/.minemacs.d/.emacs.d/snippets`" instead?
So, "`(user-emacs-directory)/snippets`" instead of "`$MINEMACSDIR/snippets`".

**Comment by abougouffa** [comment]
I wonder why you still have the `~/.emacs.d/snippets/` directory! I cannot reproduce it on my machine!

For the second point, in fact, MinEmacs (a bit like Doom or Spacemacs) should reside in `~/.emacs.d` (or any other equivalent directory), and the `~/.minemacs.d/` (a bit like `~/.doom.d/` or `~/.spacemacs.d/`) should contain user config files (`config.el`, `modules.el`, `early-config.el`, ...). The `snippets` directory is supposed to include the user defined snippets, so logically it should reside in the user config directory.

**Comment by JacekHoleczek** [comment]
I suspect that the described "`snippets`" directory problems are "interconnected".

I tried the "canonical" way of installing [MinEmacs](https://github.com/abougouffa/minemacs):

```bash
rm -rf ~/.emacs.el ~/.emacs ~/.emacs.d
rm -rf ~/.minemacs.d
git clone --depth 1 https://github.com/abougouffa/minemacs.git ~/.emacs.d
emacs
```

Quite early in the "bootstrap", it creates the empty "`~/.emacs.d/snippets`" directory, either when processing the "`yasnippet-snippets`" and the "`doom-snippets`" (judging from the printed output in the "echo area" line in the bottom of the emacs frame, but I don't know how much "delay" this printout has).
However, at that moment, the user config directory "`~/.minemacs.d`" is not present at all.
It is first created at the very end of the "bootstrap" (and then there immediately appears the empty "~/.minemacs.d/snippets" directory, too).

So, maybe the problem is that because the user snippets directory does not exist in "`~/.minemacs.d/`", the "bootstrap" creates a default one in "`~/.emacs.d/`".
Or maybe it simply disregards the "user config" at that moment (early in the "bootstrap").

**Comment by JacekHoleczek** [comment]
I noticed another funny quirk in the "bootstrap".

The first time I start "`emacs`", it automatically enters the "bootstrap".
I then move my "mouse focus" to work in another window(s).
The "bootstrap" continues but only until I see "`Installing parinfer-rust-linux.so v0.4.4.-beta ...`".
As soon as this line is printed (in the "echo area" line at the bottom of the emacs frame), it "hangs".
It simply waits for the "mouse focus". I need to move my mouse to the "emacs frame" and then it magically continues with downloading the "`Emoji images`" until it prints "`Extracting...done`".
Then, if I do nothing, it "hangs" again.
First when I move my "mouse focus" to another window and then back to the "emacs frame", it continues, and it finally creates the user config "`~/.minemacs.d/`" directory) (the "`Extracting...done`" line is still visible).

If it matters, my "`emacs`" is actually the "`emacs-lucid`", built on Ubuntu 22.04 / x86_64 / gcc 11.3.0 from an up-to-date [emacs-28](https://git.savannah.gnu.org/cgit/emacs.git/log/?h=emacs-28) branch source code.

**Comment by abougouffa** [comment]
Hello @JacekHoleczek

Thanks again for the feedback,

I refactored yesterday's code by adding a way to ensure (in most of the cases) that the directories gets created in the right place (in `minemacs-local-dir`).

For the focus issue, I've noticed it as well, but I didn't dig deeper since it is not a fatal issue.

It happens when loading lazy packages (loaded after startup when Emacs goes idle), it seems that Emacs' `run-with-idle-timer` behaves strangely when called in such a context. I will try to investigate this when I get some free time.

In the meanwhile, I think with la latest commits, all the blocking issues are fixed.

**Comment by JacekHoleczek** [comment]
I've redone the "canonical" test. Before starting "`emacs`", I first ran "`mkdir -p ~/.minemacs.d/snippets`". It didn't help. It still created the "`~/.emacs.d/snippets`" (at the same time as previously).
So, it seems it disregards the "user config" at that moment (early in the "bootstrap").

**Comment by abougouffa** [comment]
I've tested it on a VM with Emacs 28 on Ubuntu, I got the same effect! The snippets' directory is created at startup. I tried to force `yasnippet` (which is causing the creation of the directory) to use another path, but it still creates it (when byte-compiling the package). 

I don't understand why it is created even when the variable is overwritten!

I will try to take a look at it when I get some free time.
